howso/ablation.amlg

```diff
@@ -109,6 +109,229 @@
+								))
+							)
+							;(current_value) is the numerical value of the threshold. Exceeding this should satisfy the
+							; threshold and stop reduction.
+							(current_value)
+						)
+					)
+					;(current_value) contains the threshold map for a specific measure, e.g. for accuracy {target 0.05}
+					; would indicate a % change of greater than 5% would satisfy the threshold and stop reduction.
+					(current_value)
+				)
+			)
+			rel_threshold_map
+		)
+	)
+
+	;helper method which flattens the result of one of the !Check*Thresholds methods
+	; to determine whether at least one of the thresholds was violated (true) or not (false).
+	#!FlattenThresholdResultMap
+	(declare
+		(assoc
+			threshold_result_map (assoc)
+		)
+		(apply "or"
+			(values (map
+				(lambda
+					(apply "or" (values (current_value)))
+				)
+				threshold_result_map
+			))
+		)
+	)
+
+	;check all of the threshold maps against the current prediction stats map or the current
+	; and previous prediction stats maps. Returns the threshold map that failed.
+	;
+	;parameters:
+	; abs_threshold_map: a map of a measure to a map of feature name to absolute score threshold
+	; delta_threshold_map: a map of measure to a map of feature name to delta score threshold
+	; rel_threshold_map: a map of measure to a map of feature name to relative score threshold
+	; prev_prediction_stats_map: the previous prediction stats map.
+	; new_prediction_stats_map: the new prediction stats map.
+	#!CheckThresholds
+	(declare
+		(assoc
+			abs_threshold_map (assoc)
+			delta_threshold_map (assoc)
+			rel_threshold_map (assoc)
+			prev_prediction_stats_map (assoc)
+			new_prediction_stats_map (assoc)
+
+			;internal
+			abs_threshold_result_map (null)
+			delta_threshold_result_map (null)
+			output (assoc)
+		)
+
+		(assign (assoc
+			abs_threshold_result_map
+				(call !CheckAbsThresholds (assoc
+					abs_threshold_map abs_threshold_map
+					prediction_stats_map new_prediction_stats_map
+				))
+			delta_threshold_result_map
+				(call !CheckDeltaThresholds (assoc
+					delta_threshold_map delta_threshold_map
+					prev_prediction_stats_map prev_prediction_stats_map
+					new_prediction_stats_map new_prediction_stats_map
+				))
+			rel_threshold_result_map
+				(call !CheckRelThresholds (assoc
+					rel_threshold_map rel_threshold_map
+					prev_prediction_stats_map prev_prediction_stats_map
+					new_prediction_stats_map new_prediction_stats_map
+				))
+		))
+
+		(if (call !FlattenThresholdResultMap (assoc threshold_result_map abs_threshold_result_map))
+			(accum "output" ["abs"] abs_threshold_result_map)
+		)
+		(if (call !FlattenThresholdResultMap (assoc threshold_result_map delta_threshold_result_map))
+			(accum "output" ["delta"] delta_threshold_result_map)
+		)
+		(if (call !FlattenThresholdResultMap (assoc threshold_result_map rel_threshold_result_map))
+			(accum "output" ["rel"] rel_threshold_result_map)
+		)
+
+		output
```

Comment:
Maybe rather than (accum)'s to the empty (assoc) and returning it at the end, we could just build it at the end with something like:

```
(append
    (if (call !Flatten... )
        {"abs" abs_thres...}
        {}
    )
    (if (call !Flatten... )
        {"delta" delta_thres...}
        {}
    )
    (if (call !Flatten... )
        {"rel" rel_thres...}
        {}
    )
)
```

might be a little uglier to some, but would save us some copies?